zppy/templates/coupled_global.py

```diff
@@ -662,210 +1205,194 @@ def make_plot_pdfs(
+    # shutil.rmtree("table")
+    # new_url = f"viewer_{component}"
+    # # shutil.move("viewer", new_url)
+    # distutils.dir_util.copy_tree("viewer", new_url)
+    return url
+
+
+# Copied from E3SM Diags and modified
+def create_viewer_index(
+    root_dir: str, title_and_url_list: List[Tuple[str, str]]
+) -> str:
+    """
+    Creates the index page in root_dir which
+    joins the individual viewers.
+
+    Each tuple is on its own row.
+    """
+
+    def insert_data_in_row(row_obj, name, url):
+        """
+        Given a row object, insert the name and url.
+        """
+        td = soup.new_tag("td")
+        a = soup.new_tag("a")
+        a["href"] = url
+        a.string = name
+        td.append(a)
+        row_obj.append(td)
+
+    install_path = ""  # TODO: figure this out
+    path = os.path.join(install_path, "viewer", "index_template.html")
+    output = os.path.join(root_dir, "index.html")
+
+    soup = BeautifulSoup(open(path), "lxml")
+
+    # If no one changes it, the template only has
+    # one element in the find command below.
+    table = soup.find_all("table", {"class": "table"})[0]
+
+    # Adding the title.
+    tr = soup.new_tag("tr")
+    th = soup.new_tag("th")
+    th.string = "Output Sets"
+    tr.append(th)
+
+    # Adding each of the rows.
+    for row in title_and_url_list:
+        tr = soup.new_tag("tr")
+
+        if isinstance(row, list):
+            for elt in row:
+                name, url = elt
+                insert_data_in_row(tr, name, url)
         else:
-            raise RuntimeError(f"Invalid rgn={rgn}")
-        run(parameters, rgn)
+            name, url = row
+            insert_data_in_row(tr, name, url)
+
+        table.append(tr)
+
+    html = soup.prettify("utf-8")
+
+    with open(output, "wb") as f:
+        f.write(html)
+
+    return output
+
+
+def run_by_region(command_line_arguments):
+    parameters: Parameters = Parameters(command_line_arguments)
+    requested_variables = RequestedVariables(parameters)
+    for rgn in parameters.regions:
+        run(parameters, requested_variables, rgn)
+    plots_per_page = parameters.nrows * parameters.ncols
+    # TODO: Is this how we want to determine when to make a viewer or should we have a `make_viewer` parameter in the cfg?
+    if plots_per_page == 1:
+        # In this case, we don't want the summary PDF.
+        # Rather, we want to construct a viewer similar to E3SM Diags.
+        # TODO: determine directory paths for each viewer
+        # TODO: include "original"?
+        # for component in ["original", "atm", "ice", "lnd", "ocn"]:
+        title_and_url_list: List[Tuple[str, str]] = []
+        for component in ["lnd"]:
+            vars = get_vars(requested_variables, component)
+            url = create_viewer(parameters, vars, component)
+            print(url)
+            title_and_url_list.append((component, url))
+        # index_url: str = create_viewer_index(parameters.case_dir, title_and_url_list)
```

Comment:
@chengzhuzhang I will work on this more, but I'm having some difficulties with setting up multiple viewers (see earlier comment) and then getting them to show up on an index page together. I think I just need to match up file paths and URLs correctly. 

The problem is `viewer` seems to be used everywhere, so it's difficult to differentiate between viewers. For instance, in https://github.com/E3SM-Project/e3sm_diags/blob/7b054251e54f223e44ec1e15908548e8e1797744/e3sm_diags/viewer/index_template.html, the code block
```
  <link href="viewer/css/bootstrap.min.css" rel="stylesheet" type="text/css"/>
  <link href="viewer/css/viewer.css" rel="stylesheet" type="text/css"/>
  <script src="viewer/js/jquery-2.2.3.min.js" type="text/javascript">
  </script>
  <script src="viewer/js/bootstrap.min.js" type="text/javascript">
  </script>
  <script src="viewer/js/viewer.js" type="text/javascript">
```
implies the existence of a `viewer` directory above all the other viewers.

Again, I'll look into this more myself, but I am wondering if we could get the multi-viewer index working quite quickly if we just pair programmed it together.

Reply:
I'm not sure about issues here. Is the idea to create a main viewer and have each component viewers link to it? As a first step, does it work to create each component viewer under each directory?

Reply:
**What I'm aiming for:**

Something like [e3sm_diags](https://web.lcrc.anl.gov/public/e3sm/zppy_test_resources/expected_comprehensive_v3/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1987-1988/viewer/) where there's an index page linking to multiple viewers:
![Screenshot 2024-10-25 at 9 57 29 AM](https://github.com/user-attachments/assets/b0c8da15-6a40-4efe-8767-a27e7c0947e9)

I want a viewer for each component.

**What I currently have:**

[My latest results](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_viewers_www/test-pr616-20241023v9/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/viewer/) are just the land plots:
![Screenshot 2024-10-25 at 10 03 09 AM](https://github.com/user-attachments/assets/d0b97913-bb4f-46d7-adb9-c558c7f8f68c)


The [parent directory](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_viewers_www/test-pr616-20241023v9/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/) has a subdir `viewer/`.

If I try to create multiple viewers by having two subdirs `viewer_lnd` and `viewer_atm`, then the formatting goes off, as described in https://github.com/E3SM-Project/zppy/pull/616#discussion_r1814065066. This can be seen [here](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_viewers_www/test-pr616-20241023v8/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/viewer_lnd/):
![Screenshot 2024-10-25 at 10 02 33 AM](https://github.com/user-attachments/assets/17f5cddc-2c7a-43c6-8f26-0d892343a254)
and [here](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_min_case_global_time_series_viewers_www/test-pr616-20241023v8/v3.LR.historical_0051/global_time_series/global_time_series_1985-1995_results/viewer_atm/):
![Screenshot 2024-10-25 at 10 04 43 AM](https://github.com/user-attachments/assets/5099079f-f351-4ee9-824e-ed1a109030d3)
Actually, `atm` is even worse; the plots don't seem to get linked from the `atm` viewer.

Reply:
@tomvothecoder The issues I mentioned with the Viewer generation are described in-detail above.

Reply:
It seems like you're on the right track by generating one viewer sub-directory per component. And as you mentioned, there might be issues with relative links. It looks like each viewer's HTML is not referencing the `/viewer/css` and `/viewer/js` files, resulting in messed up formatting. I'm guessing the same issue is happening with the links to the `atm` plots.

---
For context, this is how the E3SM diagnostics viewer is structured and operates.

Directory structure
```html
/viewer -- root dir
  /viewer -- main viewer dir
    /css. -- CSS dir shared across each set viewer
    /js. - JavaScript dir, shared across each set viewer
  index.html  -- main index page that links to each set viewer (e.g., `lat_lon`, `polar`)
  /lat_lon
    index.html. -- lat lon viewer index oage
    /aod_550 -- variable
        /aoddust-global-macv2 
          ann.html
          jja.html
   /polar
     ...
```

1. Generate a base files:
   - `/viewer/viewer` dir that contains `/css` and `/js` for CSS and JavaScript files 
   -  `index.html` which is the [index page](https://web.lcrc.anl.gov/public/e3sm/zppy_test_resources/expected_comprehensive_v3/e3sm_diags/atm_monthly_180x360_aave/model_vs_obs_1987-1988/viewer/) you linked above. This file provides links to each set's viewer `index.html`, example below.
```html
	<!--
	    A template used to create the index.
	    This index has links to all of the separate viewers.
	-->
	<!DOCTYPE html>
	<html>
	 <head>
	  <title>
	   E3SM Diagnostics
	  </title>
	  <meta content="width=device-width, intial-scale=1" name="viewport"/>
	  <meta charset="utf-8"/>
	  <link href="viewer/css/bootstrap.min.css" rel="stylesheet" type="text/css"/>
	  <link href="viewer/css/viewer.css" rel="stylesheet" type="text/css"/>
	  <script src="viewer/js/jquery-2.2.3.min.js" type="text/javascript">
	  </script>
	  <script src="viewer/js/bootstrap.min.js" type="text/javascript">
	  </script>
	  <script src="viewer/js/viewer.js" type="text/javascript">
	  </script>
	 </head>
	 <body>
	  <div id="e3sm-header" style="background-color:#dbe6c5; float:left; width:45%">
	   <p style="margin-left:5em">
	    <b>
	     E3SM Diagnostics Package v2.12.0
	     <br/>
	    </b>
	    Test: e3sm_v2
	    <br/>
	    Reference: Observation and Reanalysis
	    <br/>
	    Created: 2024-09-26 16:58:04
	   </p>
	  </div>
	  <div id="e3sm-header2" style="background-color:#dbe6c5; float:right; width:55%">
	   <img alt="logo" src="./viewer/e3sm_logo.png" style="width:201px; height:91px; background-color:#dbe6c5"/>
	  </div>
	  <div class="container">
	   <div class="row">
	    <div class="col-sm-5 col-sm-offset-1">
	     <table class="table">
	      <!--The rows get inserted here.-->
	      <tr>
	       <td>
	        <a href="lat_lon/index.html">
	         Latitude-Longitude contour maps
	        </a>
	       </td>
	       <td>
	        <a href="table/index.html">
	         Table
	        </a>
	       </td>
	       <td>
	        <a href="taylor/index.html">
	         Taylor Diagram
	        </a>
	       </td>
	       <td>
	        <a href="cmip6/index.html">
	         CMIP6 Comparison
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="zonal_mean_xy/index.html">
	         Zonal mean line plots
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="zonal_mean_2d/index.html">
	         Pressure-Latitude zonal mean contour plots
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="zonal_mean_2d_stratosphere/index.html">
	         Pressure-Latitude zonal mean contour plots (Stratosphere)
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="polar/index.html">
	         Polar contour maps
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="cosp_histogram/index.html">
	         CloudTopHeight-Tau joint histograms
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="meridional_mean_2d/index.html">
	         Pressure-Longitude meridional mean contour plots
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="annual_cycle_zonal_mean/index.html">
	         Annual Cycle Zonal Mean Contour Plots
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="enso_diags/index.html">
	         ENSO Diagnostics
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="qbo/index.html">
	         Quasi-biennial Oscillation
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="area_mean_time_series/index.html">
	         Area Mean Time Series
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="diurnal_cycle/index.html">
	         Diurnal cycle phase maps
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="streamflow/index.html">
	         Streamflow
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="arm_diags/index.html">
	         Diagnostics at ARM stations
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="tc_analysis/index.html">
	         Diagnostics for Tropical Cyclones
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="aerosol_aeronet/index.html">
	         Aerosol Diags at AERONET sites
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="aerosol/index.html">
	         Aerosol Budget Tables
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="mp_partition/index.html">
	         Mixed-phase Partition
	        </a>
	       </td>
	      </tr>
	      <tr>
	       <td>
	        <a href="../prov">
	         Provenance
	        </a>
	       </td>
	      </tr>
	     </table>
	    </div>
	    <div class="col-sm-5">
	     <div class="img_links">
	     </div>
	    </div>
	   </div>
	  </div>
	 </body>
	</html>
```
2. Generate viewer for each set and store in sub-directories (e.g., `/viewer/lat_lon`, `/viewer/polar`)
   - Contains an `index.html`
   - Contains sub-directories with `.html` for each page
   - These HTML Files reference the base `/viewer/viewer` CSS and JS files